zkml/src/layers/flatten.rs

```diff
@@ -0,0 +1,100 @@
+use ff_ext::ExtensionField;
+use serde::{Deserialize, Serialize, de::DeserializeOwned};
+
+use crate::{
+    NextPowerOfTwo, Tensor,
+    commit::precommit::PolyID,
+    iop::context::ContextAux,
+    layers::LayerCtx,
+    padding::{PaddingMode, ShapeInfo, reshape},
+    tensor::Number,
+};
+
+use super::provable::{Evaluate, LayerOut, Op, OpInfo, PadOp, ProvableOp, ProvableOpError, ProveInfo};
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Flatten;
+
+impl OpInfo for Flatten {
+    fn output_shapes(
+        &self,
+        input_shapes: &[Vec<usize>],
+        _padding_mode: PaddingMode,
+    ) -> Vec<Vec<usize>> {
+        input_shapes
+            .into_iter()
+            .map(|s| vec![s.iter().product()])
+            .collect()
+    }
+
+    fn num_outputs(&self, num_inputs: usize) -> usize {
+        num_inputs
+    }
+
+    fn describe(&self) -> String {
+        "Reshape".to_string()
+    }
+
+    fn is_provable(&self) -> bool {
+        false
+    }
+}
+
+impl<N: Number> Evaluate<N> for Flatten {
+    fn evaluate<E: ExtensionField>(
+        &self,
+        inputs: &[&Tensor<N>],
+        _unpadded_input_shapes: Vec<Vec<usize>>,
+    ) -> Result<LayerOut<N, E>, super::provable::ProvableOpError> {
+        if inputs.len() != 1 {
+            return Err(super::provable::ProvableOpError::ParameterError(
+                "Reshape expects exactly one input".to_string(),
+            ));
+        };
+        let input = inputs[0];
+        Ok(LayerOut::from_vec(vec![input.flatten()]))
+    }
+}
+
+impl<E> ProveInfo<E> for Flatten
+where
+    E: ExtensionField + DeserializeOwned,
+    E::BaseField: Serialize + DeserializeOwned,
+{
+    fn step_info(
+        &self,
+        _id: PolyID,
+        mut aux: ContextAux,
+    ) -> Result<(LayerCtx<E>, ContextAux), ProvableOpError> {
+        aux.last_output_shape
+            .iter_mut()
+            .for_each(|s| *s = s.next_power_of_two());
+        Ok((LayerCtx::Flatten, aux))
+    }
+}
+
+impl PadOp for Flatten {
+    fn pad_node(self, si: &mut ShapeInfo) -> Result<Self, super::provable::ProvableOpError>
+    where
+        Self: Sized,
+    {
+        reshape(si).map_err(|e| ProvableOpError::GenericError(e))
+    }
+}
+
+impl<E> ProvableOp<E> for Flatten
+where
+    E: ExtensionField,
+    E::BaseField: Serialize + DeserializeOwned,
+    E: Serialize + DeserializeOwned,
+{
+    type Ctx = LayerCtx<E>; // Unused
+}
```

Comment:
Not something to act on now, but it feels to me we could get rid of the necessity of implementing `ProvableOp` for ops that are not provable.
SInce we know if an op is provable or not 
```
    fn is_provable(&self) -> bool {
        false
    }
```
then we can simply try to cast via `as` only if the layer is provable, no ? is there something preventing us to do this ?

Reply:
Well, I understand the concern of simplicity but it's just weird semantic here that we are indeed implementing a provable op for an operation that is not provable. Understanding why is not that simple anymore. 
Would it be hard to do what you suggest ? It doesn't seem like it but you have the full context.

Reply:
It's not hard, it's just the code in `Layer` that will become a bit more verbose

Reply:
Done in commit https://github.com/Lagrange-Labs/deep-prove/pull/114/commits/fa37beb0683993623225ff7b66e3fbfa99564baf